[WIP] - Upgrade to Roslyn packages 1.0.0 rc2

filipw opened the pull request:
Opening the PR for visibility and potential discussions.
This PR upgrade old Roslyn CTP from September 2012 to version 1.0.0 RC2 (.NET 4.5 compatible). Also includes a VB engine. The core work was done in this repo already https://github.com/scriptcs/scriptcs-engines

Version 1.1.0-beta is .NET 4.6 only so we will have to discuss separately what to do about it.

## TODO

 - [x] fix/troubleshoot the 8 failing tests
 - [x] internalize the Roslyn DLLs into a lib folder (currently referenced from NuGet as `pre` packages)
 - [x] ensure the new engines actually run (they used to work as external modules, but now due to changes around `ILog`, they don't)
 - [x] fix `build.cmd` (*note: acceptance tests still fail*)
 - [x] fix nuget package creation (*note: see questions below, we now have new nuget packages: `scriptcs.csharp` and `scriptcs.visualbasic`)
 - [x] fix acceptance tests
 - [x] check build on OS X (old solution used to build there)
 - [x] fix Travis CI (although the issue on Travis existed before see https://travis-ci.org/scriptcs/scriptcs/builds/68478618
 - [ ] smoke tests etc etc

## Questions & issues

 - What do we do about naming? In the engines repo we called it `ScriptCs.CSharp`, should we keep this name or stick with `ScriptCs.Engine.Roslyn`? Right now I renamed everything to `csharp` and `visualbasic` respectively, so `roslyn` is no longer visible anywhere

 - This bug https://github.com/dotnet/roslyn/issues/1012 forced me to comment out 2 tests, and we no longer can determine if an invalid namespace has been added. It is a regression compared to old Roslyn engine.

 - How do we determine that this should be merged? I would like to establish some process that will allow us to get this moving, there is no point in waiting and waiting

## Next steps

Update to Roslyn 1.1.0-beta? However, as mentioned, that's .NET 4.6 only.

filipw commented:
@adamralph can you fix acceptance tests here? they never run on my machine (also in the old solution)

filipw commented:
resolved conflicts following the merge of https://github.com/scriptcs/scriptcs/pull/1009

filipw commented:
w00t, green on Travis :smile:

adamralph commented:
I cherry picked the Travis-fixing commits into https://github.com/scriptcs/scriptcs/pull/1069 so that we can get in first and allow other PR's to also be green.

adamralph commented:
Should also fix https://github.com/scriptcs/scriptcs/issues/1066

adamralph commented:
Should also fix https://github.com/scriptcs/scriptcs/issues/1070

filipw commented:
I just tested both cases and yeah, they are fixed in this PR

filipw commented:
@adamralph have you made any progress on the tests?

@scriptcs/core any feedback from anyone?

filipw commented:
we need to get this moving

adamralph commented:
Agreed. This should be our top prio, it will be a huge leap forward.

I haven't had a chance to look at the acceptance tests yet. It took a lot of my time to get 0.15 out of the door and right now I need to tend to some other projects before getting onto this.

adamralph commented:
Linking to https://github.com/scriptcs/scriptcs/issues/1059

filipw commented:
@adamralph I have resolved all of the issues, except 1 test: http://teamcity.codebetter.com/viewLog.html?buildId=196969&buildTypeId=Scriptcs_CI

It works when run without `debug` flag.

When I invoke the same CSX from the command line it works in both `debug` and non-debug modes though.

filipw commented:
Here are the screenshots

with debug flag:
<img width="1381" alt="screenshot 2015-08-12 01 10 16" src="https://cloud.githubusercontent.com/assets/1710369/9213068/f852a64e-408e-11e5-8100-42fda6eb7fdc.png">

without it:
<img width="1376" alt="screenshot 2015-08-12 01 10 29" src="https://cloud.githubusercontent.com/assets/1710369/9213076/0447292a-408f-11e5-9447-dbf020044cab.png">

filipw commented:
I have changed the one offending test to run without debug mode and now it passes.
Given the two screenshots I posted above (everything works correctly from command line) it really looks like a false positive and bug must be somewhere in the test set up.

Anyway, all is green now, so maybe we can do the usual smoke test and get this moving.

filipw commented:
I added breaking-api cause the names of the classes changed, see the "Questions" in the PR

adamralph commented on src/ScriptCs.CSharp/CSharpScriptCompilerEngine.cs:
```diff
@@ -2,33 +2,23 @@
-            Guard.AgainstNullArgument("logProvider", logProvider);
-
-            _log = logProvider.ForCurrentType();
-
```
I'm not a fan of this change. I don't think the base should expose it's `ILog`. The `ILogProvider` gets passed up the constructor call stack and each class can harvest it's own `ILog` from it. If derived classes use an `ILog` from a base class then the logger written when running under `-debug` will be the same for all the classes in the inheritance graph which is quite a loss of information.

filipw replied:
good suggestion, agree

adamralph commented:
I keep forgetting - does the engine based on Roslyn 1.0.0 rc2 run on Mono on *nix?

adamralph commented on test/ScriptCs.Tests.Acceptance/ScriptLibraries.cs:
```diff
@@ -72,7 +72,7 @@ public static void LoadingFromANonRootedScript(ScenarioDirectory directory, Scen
 
             "When executing the script"
-                .f(() => output = ScriptCsExe.Run(Path.Combine("script", "foo.csx"), directory));
+                .f(() => output = ScriptCsExe.Run(Path.Combine("script", "foo.csx"), false, directory));
```
I guess this mean the execution of scripts which use script libraries fails when running under `-debug`?

filipw replied:
no, it only fails from the acceptance test, it works when calling from command line, I posted the screenshots before

adamralph commented on test/ScriptCs.Engine.Roslyn.Tests/CSharpScriptInMemoryEngineTests.cs:
```diff
@@ -43,12 +43,12 @@ public void ShouldExposeExceptionThrownByScriptWhenErrorOccurs()
                 var lines = new List<string>
                 {
-                    "using Sysasdasdasdtem;"
+                    "Sysasdasdasdtem;"
```
Why did this have to change?

glennblock replied:
I don't believe it supports Mono currently.

On Thu, Aug 13, 2015 at 12:00 PM Adam Ralph <notifications@github.com>
wrote:

> In test/ScriptCs.Engine.Roslyn.Tests/CSharpScriptInMemoryEngineTests.cs
> <https://github.com/scriptcs/scriptcs/pull/1064#discussion_r37012264>:
>
> >
> >                  // Arrange
> >                  var lines = new List<string>
> >                  {
> > -                    "using Sysasdasdasdtem;"
> > +                    "Sysasdasdasdtem;"
>
> Why did this have to change?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scriptcs/scriptcs/pull/1064/files#r37012264>.
>

adamralph replied:
@glennblock I guess that was in answer to https://github.com/scriptcs/scriptcs/pull/1064#issuecomment-130801874 ?

filipw replied:
@adamralph that code (invalid namespace) actually does not throw any exceptions with this version of Roslyn - this is not an issue for us, cause we don't pass namespaces into code anyway (we extract them and populate `ScriptOptions` object itself)

adamralph commented:
Since this PR adds the VB engine, I think it should include some VB acceptance tests.

adamralph commented on build/Build.proj:
```diff
@@ -43,9 +43,10 @@
   <Target Name="Test" DependsOnTargets="Build">
     <ItemGroup>
-      <TestAssemblies Include="$(Root)test\**\bin\$(Configuration)\**\*.Tests.dll;$(Root)test\**\bin\$(Configuration)\**\*.Tests.Acceptance.dll" />
+       <TestAssemblies Include="$(Root)test\**\bin\$(Configuration)\**\*.Tests.dll;$(Root)test\**\bin\$(Configuration)\**\*.Tests.Acceptance.dll" />
```
I guess lines 46 and 49 can be reverted now?

filipw replied:
yeah

adamralph commented on src/ScriptCs.Engine.Common/ScriptCs.Engine.Common.csproj:
```diff
@@ -1,18 +1,19 @@
     <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
+    <FileAlignment>512</FileAlignment>
+    <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\</SolutionDir>
+    <RestorePackages>true</RestorePackages>
```
Is this line for the old MSBuild based package restore?

adamralph commented on src/ScriptCs.Engine.Common/ScriptCs.Engine.Common.csproj:
```diff
@@ -22,7 +23,6 @@
     <DefineConstants>DEBUG;TRACE</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
-    <CodeAnalysisRuleSet>..\..\ScriptCs.ruleset</CodeAnalysisRuleSet>
```
Should this line be restored?

adamralph commented on src/ScriptCs.Engine.Common/ScriptCs.Engine.Common.csproj:
```diff
@@ -31,22 +31,31 @@
     <DefineConstants>TRACE</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
-    <CodeAnalysisRuleSet>..\..\ScriptCs.ruleset</CodeAnalysisRuleSet>
```
Should this line be restored?

adamralph commented on src/ScriptCs.Engine.Common/ScriptCs.Engine.Common.csproj:
```diff
@@ -85,13 +93,17 @@
       <Name>ScriptCs.Core</Name>
     </ProjectReference>
   </ItemGroup>
-  <Import Project="..\..\build\ScriptCs.Common.props" />
```
Should this line be restored?

filipw replied:
it's on a different line for some reason but it's there

adamralph commented on src/ScriptCs.Engine.Common/ScriptCs.Engine.Common.csproj:
```diff
@@ -85,13 +93,17 @@
   </ItemGroup>
-  <Import Project="..\..\build\ScriptCs.Common.props" />
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
-  <Import Project="..\..\packages\StyleCop.MSBuild.4.7.49.0\build\StyleCop.MSBuild.Targets" Condition="Exists('..\..\packages\StyleCop.MSBuild.4.7.49.0\build\StyleCop.MSBuild.Targets')" />
```
Should this line (and the corresponding one below) be restored?

adamralph commented on src/ScriptCs.Engine.Common/ScriptCs.Engine.Common.csproj:
```diff
@@ -85,13 +93,17 @@
+  </Target>
+  <Target Name="AfterBuild">
+  </Target>
+  -->
```
I guess this can be removed?

adamralph commented on src/ScriptCs.CSharp/ScriptCs.CSharp.nuspec:
```diff
@@ -1,20 +1,16 @@
-    <id>ScriptCs.Engine.Roslyn</id>
-    <version>$version$</version>
+    <id>ScriptCs.CSharp</id>
+    <!-- The version will be replaced by the build script-->
```
I think this was only true over in the engines repo, since that used Bau.

adamralph commented on src/ScriptCs.CSharp/CSharpModule.cs:
```diff
@@ -1,9 +1,10 @@
-﻿using ScriptCs.Contracts;
+﻿using System;
```
Seems redundant

filipw commented:
@adamralph I have addressed all of your feedback

Regarding VB I suggest we don't spend time on it now, let's just get this thing out of the door. We can simply just not advertise VB for now (or even make it no-op for now)

filipw commented:
can we move forward with this?

I have been keeping this branch up to date for over 2 months now and I see no reason for us to stay behind on this. The only issue I can see, and I want everyone to be every of is that I renamed the roslyn module to `Scriptcs.CSharp` like we had it in the other engine repo. This also means a nuget package name change.

adamralph commented:
Should also fix #1095

jrusbatch commented:
:+1:

jrusbatch commented:
:clap:

filipw commented:
![](https://media4.giphy.com/media/czoS1CAP2YZBS/200.gif)

glennblock commented:
Rockin!

On Tue, Sep 29, 2015 at 12:55 PM Filip W <notifications@github.com> wrote:

>
> <https://camo.githubusercontent.com/5f183850601dcd1b0087ee70b2d30410c9af4d84/68747470733a2f2f6d65646961342e67697068792e636f6d2f6d656469612f637a6f533143415032595a42532f3230302e676966>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scriptcs/scriptcs/pull/1064#issuecomment-144170425>.
>

adamralph commented:
re #1097
